cmd/testvectors/main.go

```diff
@@ -0,0 +1,272 @@
+			typeTx:      Spawn,
+		},
+	}
+}
+
+func generateTestVectors() []TestVector {
+	// we only need one keypair
+	pub, priv := getKey()
+	pubCore := types.BytesToHash(pub.Bytes())
+
+	// read network configs - needed for genesisID
+	var configMainnet, configTestnet config.GenesisConfig
+	configMainnet = config.MainnetConfig().Genesis
+	if testnet, err := presets.Get("testnet"); err != nil {
+		logrus.Fatalf("Error getting testnet config: %v", err)
+	} else {
+		configTestnet = testnet.Genesis
+	}
+	// not sure how to get hrp programmatically from config so we just hardcode it
+	networks := map[string]config.GenesisConfig{
+		"sm":    configMainnet,
+		"stest": configTestnet,
+	}
+
+	testVectors := make([]TestVector, 0)
+	index := 0
+	nonce := uint64(0)
+	amount := uint64(constants.OneSmesh)
+	// just use a single, random destination address
+	// note: destination is not used in all tx types
+	destination := generateAddress()
+	for hrp, netconf := range networks {
+		// hrp is used in address generation
+		types.SetNetworkHRP(hrp)
+
+		// initialization
+		genesisID := netconf.GenesisID()
+		opts := []sdk.Opt{
+			sdk.WithGenesisID(genesisID),
+			sdk.WithGasPrice(1),
+		}
+
+		// we need a VM object for validation and gas cost computation
+		vm := genvm.New(sql.InMemory(), genvm.WithConfig(genvm.Config{GasLimit: math.MaxUint64, GenesisID: genesisID}))
+
+		// SIMPLE WALLET (SINGLE SIG)
+		spawnArgsWallet := &templateWallet.SpawnArguments{
+			PublicKey: pubCore,
+		}
+		principal := core.ComputePrincipal(templateWallet.TemplateAddress, spawnArgsWallet)
+
+		// our random account needs a balance so it can be spawned
+		// this is not strictly necessary for the test vectors but it allows us to perform validation
+		vm.ApplyGenesis([]types.Account{{
+			Address: principal,
+			Balance: constants.OneSmesh,
+		}})
+
+		// need a list, not a map, since order matters here
+		// (self-spawn must come before spend)
+		txList := []struct {
+			txtype typeTx
+			tx     []byte
+		}{
+			{txtype: Spawn, tx: sdkWallet.Spawn(priv, templateWallet.TemplateAddress, spawnArgsWallet, nonce, opts...)},
+			{txtype: SelfSpawn, tx: sdkWallet.SelfSpawn(priv, nonce, opts...)},
+			{txtype: Spend, tx: sdkWallet.Spend(priv, destination, amount, nonce, opts...)},
+		}
+		for _, txPair := range txList {
+			logrus.Debugf("[%d] Generating test vector for %s %s %s", index, hrp, "wallet", txPair.txtype)
+			testVector := txToTestVector(txPair.tx, vm, index, amount, nonce, Wallet, txPair.txtype, destination.String(), hrp)
+			testVectors = append(testVectors, testVector)
+			index++
+		}
+
+		// MULTISIG
+		// include 1- and 2- of -1 and -2 (m of n)
+		for _, m := range []uint8{1, 2} {
+			for _, n := range []uint8{1, 2} {
+				// fill in the missing public keys (we already have one)
+				pubKeys := make([]core.PublicKey, n)
+
+				// frustratingly, we need the same list of pubkeys in a different format
+				// https://github.com/spacemeshos/go-spacemesh/issues/6061
```

Comment:
It is not a different format - `signing.Pubkey` is a wrapper around `ed25519.Pubkey` to add additional methods. The conversion to bytes and then `ed25519.Pubkey` isn't necessary since the later is a field in the wrapper:

```go
edPubKeys := make([]ed25519.PublicKey, n)
edPubKeys[0] = pub.PublicKey
```

The real problem is `core.PublicKey` why is it defined as `types.Hash32`? A key isn't a hash, and `types.Hash32` is an array, not a slice so every conversion between a "real" key and this custom type requires a copy of the value.

This line:

```go
core.PublicKey(types.BytesToHash(pub2.Bytes()))
```

copies the key to an array and then the function using the `core.PubicKey` copies it back into a slice when using it 😕